packages/ses/src/make-hardener.js

```diff
@@ -218,21 +220,53 @@ export const makeHardener = () => {
             enqueue(desc.set, `${pathname}(set)`);
           }
         });
-      }
+      };
+
+      const freezeAndTraverse =
+        FERAL_STACK_GETTER === undefined
+          ? // On platforms without v8's error own stack accessor problem,
+            // don't pay for any extra overhead.
+            baseFreezeAndTraverse
+          : obj => {
+              if (isError(obj)) {
+                // Only pay the overhead if it first passes this cheap isError
+                // check. Otherwise, it will be unrepaired, but won't be judged
+                // to be a passable error anyway, so will not be unsafe.
+                const stackDesc = getOwnPropertyDescriptor(obj, 'stack');
+                if (
+                  stackDesc &&
+                  stackDesc.get === FERAL_STACK_GETTER &&
+                  stackDesc.configurable
+                ) {
+                  // Can only repair if it is configurable. Otherwise, leave
+                  // unrepaired, in which case it will not be judged passable,
+                  // avoiding a safety problem.
+                  defineProperty(obj, 'stack', {
+                    // NOTE: Calls getter during harden, which seems dangerous.
+                    // But we're only calling the problematic getter whose
+                    // hazards we think we understand.
+                    // @ts-expect-error TS should know FERAL_STACK_GETTER
+                    // cannot be `undefined` here.
```

Comment:
Yeah type narrowing does not cross function boundaries in general.

Reply:
> Yeah type narrowing does not cross function boundaries in general.

Any idea why not?

Reply:
Most likely because most people do not define functions conditionally, so it's a case not worth handling for the compiler.

While not the case here because it's a `const` (which TS might not track, I don't know), there are a lot of cases where TypeScript "resets" type inference at function call boundaries because it cannot know whether the function is immediately invoked, or if other calls may have mutated the narrowed variable.

Reply:
Perhaps the type inference doesn't distinguish between the temporal implications of `let` vs `const`? If `FERAL_STACK_GETTER` were a `let`, then, since a function may be invoked later, it might be invoked well after the `FERAL_STACK_GETTER === undefined` conditioned block. However, `const` guarantees stability, and therefore the absence of these temporal hazards. Does this sound right? (Attn @turadg @michaelfig ) Is this something they should fix? If so, worth filing an issue?

Reply:
Plenty of reasons this might not maintain the narrowing, but given how edgy this case is, I don't believe it's worth getting TS to improve support for this. There is also a work around if one can avoid a ternary: assign to a `const` after narrowing.

Reply:
Thanks.

I added a link to this PR conversation, which I think is adequate for this PR. Resolving.